libraries/AC_AutoTune/AC_AutoTune.cpp

```diff
@@ -1713,56 +1290,831 @@ void AC_AutoTune::get_poshold_attitude(float &roll_cd_out, float &pitch_cd_out,
-// @Field: RP: new rate gain P term
-// @Field: RD: new rate gain D term
-// @Field: SP: new angle P term
-// @Field: ddt: maximum measured twitching acceleration
-
-// Write an Autotune data packet
-void AC_AutoTune::Log_Write_AutoTune(uint8_t _axis, uint8_t tune_step, float meas_target, float meas_min, float meas_max, float new_gain_rp, float new_gain_rd, float new_gain_sp, float new_ddt)
+void AC_AutoTune::twitch_test_init()
+{
+    float target_max_rate;
+    switch (axis) {
+    case ROLL:
+        target_max_rate = MAX(AUTOTUNE_TARGET_MIN_RATE_RLLPIT_CDS, step_scaler*AUTOTUNE_TARGET_RATE_RLLPIT_CDS);
+        target_rate = constrain_float(ToDeg(attitude_control->max_rate_step_bf_roll())*100.0f, AUTOTUNE_TARGET_MIN_RATE_RLLPIT_CDS, target_max_rate);
+        target_angle = constrain_float(ToDeg(attitude_control->max_angle_step_bf_roll())*100.0f, AUTOTUNE_TARGET_MIN_ANGLE_RLLPIT_CD, AUTOTUNE_TARGET_ANGLE_RLLPIT_CD);
+        rotation_rate_filt.set_cutoff_frequency(attitude_control->get_rate_roll_pid().filt_D_hz()*2.0f);
+        break;
+    case PITCH:
+        target_max_rate = MAX(AUTOTUNE_TARGET_MIN_RATE_RLLPIT_CDS, step_scaler*AUTOTUNE_TARGET_RATE_RLLPIT_CDS);
+        target_rate = constrain_float(ToDeg(attitude_control->max_rate_step_bf_pitch())*100.0f, AUTOTUNE_TARGET_MIN_RATE_RLLPIT_CDS, target_max_rate);
+        target_angle = constrain_float(ToDeg(attitude_control->max_angle_step_bf_pitch())*100.0f, AUTOTUNE_TARGET_MIN_ANGLE_RLLPIT_CD, AUTOTUNE_TARGET_ANGLE_RLLPIT_CD);
+        rotation_rate_filt.set_cutoff_frequency(attitude_control->get_rate_pitch_pid().filt_D_hz()*2.0f);
+        break;
+    case YAW:
+        target_max_rate = MAX(AUTOTUNE_TARGET_MIN_RATE_RLLPIT_CDS, step_scaler*AUTOTUNE_TARGET_RATE_YAW_CDS);
+        target_rate = constrain_float(ToDeg(attitude_control->max_rate_step_bf_yaw()*0.75f)*100.0f, AUTOTUNE_TARGET_MIN_RATE_YAW_CDS, target_max_rate);
+        target_angle = constrain_float(ToDeg(attitude_control->max_angle_step_bf_yaw()*0.75f)*100.0f, AUTOTUNE_TARGET_MIN_ANGLE_YAW_CD, AUTOTUNE_TARGET_ANGLE_YAW_CD);
+        rotation_rate_filt.set_cutoff_frequency(AUTOTUNE_Y_FILT_FREQ);
+        break;
+    }
+
+    if ((tune_type == SP_DOWN) || (tune_type == SP_UP)) {
+        rotation_rate_filt.reset(start_rate);
+    } else {
+        rotation_rate_filt.reset(0);
+    }
+
+}
+
+//run twitch test
+void AC_AutoTune::twitch_test_run(AxisType test_axis, const float dir_sign)
+{
+    // disable rate limits
+    attitude_control->use_sqrt_controller(false);
+    // hold current attitude
+    attitude_control->input_rate_bf_roll_pitch_yaw(0.0f, 0.0f, 0.0f);
+
+    if ((tune_type == SP_DOWN) || (tune_type == SP_UP)) {
+        // step angle targets on first iteration
+        if (twitch_first_iter) {
+            twitch_first_iter = false;
+            // Testing increasing stabilize P gain so will set lean angle target
+            switch (test_axis) {
+            case ROLL:
+                // request roll to 20deg
```

Comment:
Pretty sure the 20 here is wrong

Reply:
What do you mean it's wrong?  This is the same code copied from above but put into a method.  This is straight from the original autotune.  we'll have to have @lthall check it.